libs/pdi-scanner/ts/src/scanner_client.ts

```diff
@@ -0,0 +1,340 @@
+  frontM4SensorCovered: boolean;
+  /**
+   * @deprecated Not used by PageScan 6, always false.
+   */
+  frontM5SensorCovered: never;
+  /**
+   * @deprecated Not used by PageScan 6, always false.
+   */
+  frontRightSensorCovered: never;
+  scannerReady: boolean;
+  xmtAborted: boolean;
+  documentJam: boolean;
+  scanArrayPixelError: boolean;
+  inDiagnosticMode: boolean;
+  documentInScanner: boolean;
+  calibrationOfUnitNeeded: boolean;
+}
+
+/**
+ * Coded error responses from the scanner client.
+ */
+export type ScannerError =
+  | { code: 'disconnected' }
+  | { code: 'alreadyConnected' }
+  | { code: 'commandInProgress' }
+  | { code: 'scanInProgress' }
+  | { code: 'scanFailed' }
+  | { code: 'other'; message: string };
+
+/**
+ * An event emitted by the scanner client *not* in response to a command. Can be
+ * received by adding a listener to the client.
+ */
+export type ScannerEvent =
+  | ({ type: 'error' } & ScannerError)
+  | { type: 'scanStart' }
+  | { type: 'scanComplete'; images: SheetOf<ImageData> };
+
+/**
+ * An event listener for any {@link ScannerEvent} emitted by the scanner.
+ */
+export type Listener = (event: ScannerEvent) => void;
+
+/**
+ * Which direction to eject the document, and whether to hold it there.
+ */
+export type EjectMotion = 'toRear' | 'toFront' | 'toFrontAndHold';
+
+/**
+ * Internal type to represent the JSON commands sent to `pdictl`
+ */
+type PdictlCommand =
+  | { type: 'exit' }
+  | { type: 'connect' }
+  | { type: 'disconnect' }
+  | { type: 'getScannerStatus' }
+  | { type: 'enableScanning' }
+  | { type: 'disableScanning' }
+  | {
+      type: 'ejectDocument';
+      ejectMotion: EjectMotion;
+    };
+
+/**
+ * Internal type to represent the JSON responses received from `pdictl`. A
+ * response can be received in response to a command, or as an unsolicited
+ * event.
+ */
+type PdictlResponse =
+  | { type: 'ok' }
+  | ({ type: 'error' } & ScannerError)
+  | { type: 'scannerStatus'; status: ScannerStatus }
+  | { type: 'scanStart' }
+  | { type: 'scanComplete'; imageData: [string, string] };
+
+type SimpleResult = Result<void, ScannerError>;
+
+function loggableResponse(response: PdictlResponse) {
+  if (response.type === 'scanComplete') {
+    return {
+      ...response,
+      imageData: response.imageData.map((imageData) => {
+        return `${imageData.length} bytes`;
+      }),
+    };
+  }
+  return response;
+}
+
+/**
+ * Creates a client for the PDI scanner. Spawns a `pdictl` process and
+ * communicates with it over stdin/stdout.
+ *
+ * A client can only be used for one lifetime of the `pdictl` process. Once the
+ * process exits and the 'disconnected' error code is returned, the client will
+ * no longer be able to send commands and a new client should be created to
+ * reconnect to the scanner.
+ */
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export function createPdiScannerClient() {
+  const pdictl = spawn(PDICTL_PATH, { detached: true });
```

Comment:
Document why `detached: true`?

Reply:
This was necessary to allow me to intercept SIGINT when I was trying to make the demo script close the scanner connection gracefully. Now that we're not worried about that anymore, it's not necessary. Removing